---
title: "PR2 Exports for EukRef""
author: "Daniel Vaulot"
date: '`r format(Sys.time(), "%d %m %Y")`'
---
# Goal
This NoteBook allows to exports specific groups of data from PR2

# Init and read PR2
Load the variables common to the different scripts and the necessary libraries

```{r init, eval=TRUE, message=FALSE, warning=FALSE}
  source('PR2_init.R', echo=FALSE)
  # source('PR2_read.R', echo=FALSE)
  # source('PR2_read_google.R', echo=FALSE)
  pr2_18S <- qs::qread(here::here("databases", "pr2_18S.qs")) 

```


```{r function_export, eval=TRUE, message=FALSE, warning=FALSE}
# Export function

pr2_export <- function(pr2,  
                       taxo_level="class",
                       taxa = c("Pelagophyceae"),
                       compress = TRUE,
                       taxo_include = FALSE
                       ){
  taxa_string = str_c(taxa, sep="_")
  
  file_head = here::here("exports", taxa_string, "pr2_export")
  
  try(fs::dir_create(here::here("exports",taxa_string)))
  
  taxo_level_bayes = str_c(taxo_level, "_bayes")
   
  pr2_export_annotated <- pr2 %>% 
    filter(!is.na(species),
           .data[[taxo_level]] %in% taxa) %>% 
    mutate(pr2_annotated = TRUE)
  
  pr2_export_unannotated <- pr2 %>%  
    filter(is.na(species),
           .data[[taxo_level_bayes]] %in% taxa) %>% 
    mutate(pr2_annotated = FALSE) %>% 
    select(!(domain:species)) %>% 
    rename_with(~ str_replace(.x, "_bayes", ""), contains("_bayes"))
  
  pr2_export <- bind_rows(pr2_export_annotated, pr2_export_unannotated) %>% 
  # filter(!is.na(domain)) %>%   # Remove data that are not annotated yet
  mutate(species_old = species) %>% 
  arrange(domain, supergroup, division, subdivision, class, order, family, genus, species) %>% 
  relocate(species_old, .after = species) %>% 
  relocate(pr2_annotated, .after = species) %>% 
  select(-starts_with("taxo_"))
  
  pr2_export_taxonomy <- pr2_export %>% count(supergroup, division, subdivision, 
                                              class, order, family, genus, species)
  
  pr2_export_fasta_annotated <- pr2_export %>% 
    filter(!is.na(species)) %>% 
    select(seq_name=pr2_accession, supergroup:genus, species, sequence) 
  
  # pr2_export_fasta_other <- pr2_export %>% 
  #   filter(is.na(species)) %>% 
  #   mutate(seq_name=str_c(pr2_accession, gb_definition , sep = "|")) %>% 
  #   select(seq_name, sequence) 
  
  pr2_export_metadata <- pr2_export # %>%
    # select(pr2_accession:organelle, kingdom:genus, species, chimera:remark, taxon_trophic_mode:metadata_remark)
    # select(-contains("_8"), -contains("boot"))
    # select(-sequence)
    
  pr2_export_txt <- pr2_export %>% 
    mutate(tax = str_c(domain, supergroup, division, subdivision, 
                       class, order, family, genus, species, 
                       sep = "; ")) %>% 
    select(pr2_accession, tax)
  
  export(list("taxonomy"=pr2_export_taxonomy, "metadata"=pr2_export_metadata), 
         file = str_c(file_head, ".xlsx"),
         zoom=80, firstActiveRow = 2, firstActiveCol = 12, colWidths="auto",
         overwrite = TRUE
         )
  if (compress) file_ext = ".fas.gz" else file_ext = ".fas"
  dvutils::fasta_write(pr2_export_fasta_annotated, str_c(file_head, file_ext), compress, taxo_include)
  # dvutils::fasta_write(pr2_export_fasta_other, str_c(file_head, "_other", file_ext), compress, FALSE)
  
  readr::write_tsv(pr2_export_txt, str_c(file_head, ".txt"), quote = "all")
  
  cat(taxa_string, "\n")
  print(str_c("Number of sequences exported - annotated: ", nrow(pr2_export_annotated)))
  print(str_c("Number of sequences exported - other: ", nrow(pr2_export_unannotated)))
  
  return(NULL)
}
```


# List taxa of interest


```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Picozoa", "division",
  "Chlorophyta","division",
  "Amoebozoa", "supergroup",
  "Microsporida", "order",
  "Rhodophyta","division",
  "Suessiales", "order",
  "Dinophyceae", "class",
  "Metazoa", "subdivision",
  "Fungi", "subdivision"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Microsporida", "order"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Trebouxiophyceae", "class"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Ceramiales", "order",
  "Batrachospermales", "order"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Chloropicophyceae", "class"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Cnidaria", "class",
  "Ctenophora","class",
  "Placozoa","class",
  "Porifera","class"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Metchnikovellida", "order"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Eumycetozoa", "subdivision",
  "Archamoebea", "class",
  "Variosea", "class",
  "Tubulinea", "division",
  "Centramoebia", "class",
  "Flabellinia", "class",
  "Stygamoebida", "class"
)
```


```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Nucleophaga", "genus"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Apicomplexa", "subdivision"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Chrompodellids", "subdivision",
  "Alveolata_X", "subdivision",
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Chelicerata", "order"
)
```


```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Chlorodendrophyceae", "class"
)
```

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Chytridiomycota", "class"
)
```



# Export results



```{r, eval=FALSE}

  result <- purrr::map2(df$taxa, df$taxo_level, ~ pr2_export ( pr2_18S, taxa = .x, taxo_level = .y))

```
## Fungi

```{r}
pr2_fungi <- pr2_18S %>% 
  filter(!(class %in% c("Ascomycota", "Basidiomycota", "Mucoromycota")) &
         !(order %in% "Microsporida") &
         !(class_bayes %in% c("Ascomycota", "Basidiomycota", "Mucoromycota")) &
         !(order_bayes %in% "Microsporida")
           )
```

```{r}
df_fungi <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  "Fungi", "subdivision"
)
```


```{r, eval=FALSE}
  result <- purrr::map2(df_fungi$taxa, df_fungi$taxo_level, ~ pr2_export ( pr2_fungi, taxa = .x, taxo_level = .y))
```


## Chloropicophyceae

```{r}
df <- tibble::tribble(
  ~taxa, ~taxo_level,
  #--|--|----
  # "Chloropicophyceae", "class",
  "Picocystophyceae", "class"
)
```


```{r, eval=FALSE}

  result <- purrr::map2(df$taxa, df$taxo_level, 
                        ~ pr2_export ( pr2_18S, taxa = .x, taxo_level = .y, 
                                       taxo_include = TRUE, compress = FALSE))


  
```



# ==========================================================

#   NOT USED

# ==========================================================

# Taxonomy

```{r}
 pr2_taxo_class <- pr2_taxo %>% 
  group_by (kingdom, supergroup, division, class) %>% 
  tally() %>% 
   openxlsx::write.xlsx(file="../updates/2020 new taxonomy/pr2_taxonomy_class.xlsx", na="")
```


Number of sequences exported - annotated: 1336"
Number of sequences exported - other: 521"

```{r, eval=FALSE}

  target_group = "Labyrinthulomycetes"

  pr2_export <- pr2 %>% 
  filter(order == target_group |
         class_8_bayes == target_group |
         str_detect(silva_taxonomy, target_group),
         is.na(removed_version),
         !str_detect(gb_organism, "metagenome")) %>% 
  arrange(genbank_accession)

  pr2_export_xls ( pr2_export, 
                   file_head = "../exports/pr2_Labys_2023-04-03")
  
```